pkg/version/metrics.go

```diff
@@ -79,36 +82,45 @@ var (
 	}
 )
 
-func GetMetricCharts(appID string, sequence int64, prometheusAddress string) ([]MetricChart, error) {
-	if prometheusAddress == "" {
-		return []MetricChart{}, nil
+// GetGraphs returns the rendered graphs for the given app.
+// If there are no graphs or an error is encountered, the default set of graphs is returned.
+func GetGraphs(app *types.App, sequence int64, store store.Store) ([]kotsv1beta1.MetricGraph, error) {
+	graphs := DefaultMetricGraphs
+
+	archiveDir, err := ioutil.TempDir("", "kotsadm")
+	if err != nil {
+		return graphs, errors.Wrap(err, "failed to create temp dir")
 	}
+	defer os.RemoveAll(archiveDir)
 
-	db := persistence.MustGetDBSession()
-	query := `select kots_app_spec from app_version where app_id = $1 and sequence = $2`
-	row := db.QueryRow(query, appID, sequence)
+	err = store.GetAppVersionArchive(app.ID, sequence, archiveDir)
+	if err != nil {
+		return graphs, errors.Wrap(err, "failed to get app version archive")
+	}
 
-	var kotsAppSpecStr sql.NullString
-	if err := row.Scan(&kotsAppSpecStr); err != nil {
-		if err != sql.ErrNoRows {
-			return nil, errors.Wrap(err, "failed to scan")
-		}
+	kotsKinds, err := kotsutil.LoadKotsKindsFromPath(archiveDir)
+	if err != nil {
+		return graphs, errors.Wrap(err, "failed to load kots kinds from path")
 	}
 
-	graphs := DefaultMetricGraphs
-	if kotsAppSpecStr.Valid {
-		decode := scheme.Codecs.UniversalDeserializer().Decode
-		obj, _, err := decode([]byte(kotsAppSpecStr.String), nil, nil)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to decode kots app spec")
-		}
-		a := obj.(*kotsv1beta1.Application)
+	registrySettings, err := store.GetRegistryDetailsForApp(app.ID)
+	if err != nil {
+		return graphs, errors.Wrap(err, "failed to get registry settings for app")
+	}
 
-		if len(a.Spec.Graphs) > 0 {
-			graphs = a.Spec.Graphs
-		}
+	builder, err := render.NewBuilder(kotsKinds, registrySettings, app.Slug, sequence, app.IsAirgap, util.PodNamespace)
+	if err != nil {
+		return graphs, errors.Wrap(err, "failed to get template builder")
 	}
 
+	if len(kotsKinds.KotsApplication.Spec.Graphs) > 0 {
```

Comment:
the application spec should already exist in the database in the app_version table, there is no need to retrieve the archive from s3 to get it

Reply:
tip, you can use `GetDownstreamVersion` and then call `AddDetails` and the version would have kots kinds loaded from the db.

Reply:
we need more than just the application spec since we are creating a builder instance right above this.  is there a more efficient way to get the information necessary without going to s3?

Reply:
ah you're right, kots kinds are needed to render things i guess.